datafusion/expr/src/function.rs

```diff
@@ -39,6 +40,10 @@ use std::sync::Arc;
 pub type ScalarFunctionImplementation =
     Arc<dyn Fn(&[ColumnarValue]) -> Result<ColumnarValue> + Send + Sync>;
 
+/// Table function. Second tuple
+pub type TableFunctionImplementation =
+    Arc<dyn Fn(&[ColumnarValue], usize) -> Result<(ArrayRef, Vec<usize>)> + Send + Sync>;
```

Comment:
as  ArrayRef is one of ColumnarValue
``` Rust
pub enum ColumnarValue {
    /// Array of values
    Array(ArrayRef),
    /// A single value
    Scalar(ScalarValue),
}
```
I think `TableFunctionImplementation ` is same as `ScalarFunctionImplementation `.
And it only generate table N*1 , if we use as https://github.com/apache/arrow-datafusion/pull/2177#issuecomment-1100368801
```rust
Arc<dyn Fn(&[ColumnarValue], usize) -> Result<(Vec< ColumnarValue >, Vec<usize>)> + Send + Sync>;
```
We could generate N*M table
If im wrong plz correct me?

Reply:
Or in this case it can generate N*M table

Reply:
I am also a little mystified by this signature

. It looks like "Second tuple" was the start of a thought that didn't get finished? I also don't understand what the `usize` in the tuple represents -- perhaps you can add some comments explaining its purpose?

Also, I agree with @Ted-Jiang  's analysis -- I would expect this signature to return a "table" (aka a `RecordBatch` or a `Vec<ColumnarValue>` if preferred

Perhaps something like

```rust
Arc<dyn Fn(&[ColumnarValue]) -> Result<RecordBatch> + Send + Sync>;
```

or

```rust
Arc<dyn Fn(&[ColumnarValue]) -> Result<Vec<ColumnarValue>> + Send + Sync>;
```

Reply:
I guess that @gandronchik wants to chain each result(ArrayRef) of `TableFunctionImplementation` into a multi-column result (see the code in `TableFunStream::batch`). But I agree that `Arc<dyn Fn(&[ColumnarValue]) -> Result<Vec<ColumnarValue>> + Send + Sync>` is more proper.

Reply:
I updated the header of PR. Hope it is clear enough now:)